libraries/mbed/targets/hal/TARGET_ARM_SSG/TARGET_MPS2/us_ticker.c

```diff
@@ -0,0 +1,84 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <stddef.h>
+#include "us_ticker_api.h"
+#include "PeripheralNames.h"
+
+#define US_TICKER_TIMER1      CMSDK_DUALTIMER1
+#define US_TICKER_TIMER2      CMSDK_DUALTIMER2
+#define US_TICKER_TIMER_IRQn DUALTIMER_IRQn
+
+int us_ticker_inited = 0;
+
+void us_ticker_init(void) {
+    if (us_ticker_inited) return;
+    us_ticker_inited = 1;
+    
+    US_TICKER_TIMER1->TimerControl = 0x0; // disable timer
+    US_TICKER_TIMER2->TimerControl = 0x00; // disable timer
+	US_TICKER_TIMER1->TimerLoad = 0xFFFFFFFF;
+	US_TICKER_TIMER2->TimerLoad = 0xFFFFFFFF;
+	
+    US_TICKER_TIMER1->TimerControl = 0x62; // enable interrupt and set to 32 bit counter and set to periodic mode
+    US_TICKER_TIMER2->TimerControl = 0x42; // enable interrupt and set to 32 bit counter
+    
+    US_TICKER_TIMER1->TimerControl |= 0x80; // enable counter
+    US_TICKER_TIMER2->TimerControl |= 0x80; // enable counter
+    
+    NVIC_SetVector(US_TICKER_TIMER_IRQn, (uint32_t)us_ticker_irq_handler);
+    NVIC_EnableIRQ(US_TICKER_TIMER_IRQn);
+}
+
+uint32_t us_ticker_read() {
+uint32_t return_value = 0;
+    if (!us_ticker_inited)
+        us_ticker_init();
+    return_value = ((~US_TICKER_TIMER2->TimerValue)/25);
+	return return_value;
+}
+
+void us_ticker_set_interrupt(timestamp_t timestamp) {
+uint32_t timer_value = 0;
+int delta = 0;
+    if (!us_ticker_inited)
+        us_ticker_init();
+    delta = (int)(timestamp - us_ticker_read());
+    if (delta <= 0) {
+        // This event was in the past:
+        us_ticker_irq_handler();
+        return;
+    }
+		timer_value = (delta)*25;
+		// enable interrupt
+    US_TICKER_TIMER1->TimerControl = 0x0; // disable timer
+    US_TICKER_TIMER1->TimerControl = 0x62; // enable interrupt and set to 32 bit counter and set to periodic mode
+	US_TICKER_TIMER1->TimerLoad = (delta)*25; //initialise the timer value
```

Comment:
Because the divider on the timers is limited and it can’t therefore divid down to 1Mhz from the 25Mhz, so we multiply up the value for a count using a 1Mhz clock so we can use the 25Mhz clock of the timer.

From: Martin Kojtal [mailto:notifications@github.com]
Sent: 02 July 2015 15:13
To: mbedmicro/mbed
Cc: Peter Conheeney
Subject: Re: [mbed] commit of mps2 code for pull request to main mbed repo (#1219)


In libraries/mbed/targets/hal/TARGET_ARM_SSG/TARGET_MPS2/us_ticker.c<https://github.com/mbedmicro/mbed/pull/1219#discussion_r33781479>:

> +void us_ticker_set_interrupt(timestamp_t timestamp) {

> +uint32_t timer_value = 0;

> +int delta = 0;

> +    if (!us_ticker_inited)

> +        us_ticker_init();

> +    delta = (int)(timestamp - us_ticker_read());

> +    if (delta <= 0) {

> +        // This event was in the past:

> +        us_ticker_irq_handler();

> +        return;

> +    }

> +            timer_value = (delta)*25;

> +            // enable interrupt

> +    US_TICKER_TIMER1->TimerControl = 0x0; // disable timer

> +    US_TICKER_TIMER1->TimerControl = 0x62; // enable interrupt and set to 32 bit counter and set to periodic mode

> +     US_TICKER_TIMER1->TimerLoad = (delta)*25; //initialise the timer value

(delta)*25 - Why 25 ? People will ask :)

—
Reply to this email directly or view it on GitHub<https://github.com/mbedmicro/mbed/pull/1219/files#r33781479>.

-- IMPORTANT NOTICE: The contents of this email and any attachments are confidential and may also be privileged. If you are not the intended recipient, please notify the sender immediately and do not disclose the contents to any other person, use it for any purpose, or store or copy the information in any medium. Thank you.

ARM Limited, Registered office 110 Fulbourn Road, Cambridge CB1 9NJ, Registered in England & Wales, Company No: 2557590
ARM Holdings plc, Registered office 110 Fulbourn Road, Cambridge CB1 9NJ, Registered in England & Wales, Company No: 2548782